lambda_builders/builder.py

```diff
@@ -0,0 +1,129 @@
+
+
+class Builder(object):
+    """
+    Helps you build AWS Lambda functions.
+    """
+
+    # ./actions
+    _DEFAULT_BUILD_ACTIONS_DIR = os.path.join(os.path.dirname(__file__), "actions")
+
+    def __init__(self,
+                 build_actions_dir=None,
+                 entrypoint_file_name="entrypoint.py"):
+
+        """
+        Initialize the builder.
+
+        :type build_actions_dir: str
+        :param build_actions_dir:
+            Optional, path to the folder containing all the build actions. If this is not provided, then we will use
+            the default path.
+
+        :type entrypoint_file_name: str
+        :param entrypoint_file_name:
+            Optional, name of the entrypoint file that the builder should search for within each build action folder
+        """
+
+        self.build_actions_dir = build_actions_dir or self._DEFAULT_BUILD_ACTIONS_DIR
+        self.entrypoint_file_name = entrypoint_file_name
+
+    def build(self,
+              source_dir,
+              artifacts_dir,
+              manifest_path,
+              action_name,
+              runtime=None,
+              optimizations=None,
+              options=None):
+        """
+        Build a function at the given source location using the given build action.
+
+        :type source_dir: str
+        :param source_dir:
+            Path to a folder containing the source code
+
+        :type artifacts_dir: str
+        :param artifacts_dir:
+            Path to a folder where the built artifacts should be placed
+
+        :type manifest_path: str
+        :param manifest_path:
+            Path to the dependency manifest
+
+        :type action_name: str
+        :param action_name:
+            Name of the Build Action to use to build this Lambda function.
+
+        :type runtime: str
+        :param runtime:
+            Optional, name of the AWS Lambda runtime that you are building for. This is sent to the builder for
+            informational purposes.
+
+        :type optimizations: dict
+        :param optimizations:
+            Optional dictionary of optimization flags to pass to the build action. **Not supported**.
+
+        :type options: dict
+        :param options:
+            Optional dictionary of options ot pass to build action. **Not supported**.
+
+        :raises lambda_builders.exceptions.UnsupportedBuiltActionError
+        """
+
+        action_klass = self._load_action(action_name)
+        action = action_klass(source_dir,
+                              artifacts_dir,
+                              manifest_path,
+                              runtime=runtime,
+                              optimizations=optimizations,
+                              options=options)
+
+        # TODO: Verify source & artifacts directory & manifests exists
+
+        if not action.is_supported():
+            raise UnsupportedManifestError(action_name, manifest_path)
+
+        action.resolve()
```

Comment:
For python at least, part of resolving the dependencies is compiling them. Since I don't know if a particular dependency is acceptable until after I try 'compiling' it. I think this section is a noop for python.